### PR TITLE
Provide HTMLish markup alternative to Bbcode

### PR DESCRIPTION
**Pull request by daviddq**
Implements #140 

Features:

- Select your markup language of choice from the settings dialog.
- Optionally convert the already existing markup of your story to the selected format.
-index.html is fully changed since indentation was still using tabs instead of spaces.
- A bunch of code was refactored from app.js and node.js to the corresponding richTextFormatters

### CODE REVIEW

File: src/index.html [hunk 1]
```diff
@@ -1,432 +1,439 @@
...
+    <!-- ko if: (app.deleting().selected && (app.getSelectedNodes().length > 1)) -->
+    <div>Remove all selected</div>
+    <!-- /ko -->
+    <!-- ko ifnot: (app.deleting().selected) -->
+    <div>Remove <span data-bind="text: app.deleting().title"></span>?</div>
+    <!-- /ko -->
+    <!-- ko if: (app.deleting().selected && (app.getSelectedNodes().length == 1)) -->
+    <div>Remove <span data-bind="text: app.deleting().title"></span>?</div>
+    <!-- /ko -->
+    <div class="icon close" data-bind="click:app.deleting().remove;"></div>
```
Comment (blurymind) [inline src/index.html:285]: its strange we are still getting these huge difs even after we have a format command :(

**Comment by blurymind** [comment]
Something on this PR is causing the node links menu to not render  properly sometimes
Screenshot_2020-05-15_18-17-52

I have 5  nodes and none of them is listed

File: src/js/classes/ui.js [hunk 2]
```diff
@@ -71,6 +77,34 @@ export const UI = function(app) {
       .transition({ y: '-100' }, 250);
   };
 
+  // confirmMarkupConversion
+  this.confirmMarkupConversion = function () {
+    Swal.fire({
+      title: 'Are you sure?',
+      text: 'Markup on all nodes will be modified. This can rarely result in broken texts.',
```
Comment (blurymind) [inline src/js/classes/ui.js:84]: Can you also note that it is un-doable at the moment?

File: src/js/classes/ui.js [hunk 3]
```diff
@@ -71,6 +77,34 @@ export const UI = function(app) {
...
+        app.convertMarkup();
+        Swal.fire(
+          'Converted!',
+          'The markup on the nodes has been converted.',
+          'success'
+        );
+      } else if (result.dismiss === Swal.DismissReason.cancel) {
+        Swal.fire(
+          'Cancelled',
+          'The conversion has been cancelled.',
```
Comment (blurymind) [inline src/js/classes/ui.js:101]: Can we get rid of this extra cancelled pop up? The user already knows they cancelled and you are just making them click again to confirm what they already know

File: src/js/classes/richTextFormatterBbcode.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+  };
+
+  this.convert = function(text) {
+    let result = text;
+
+    result = self._convertTag('<b>(.*?)<\\/b>', '[b]${match[1]}[/b]', result);
+    result = self._convertTag('<u>(.*?)<\\/u>', '[u]${match[1]}[/u]', result);
+    result = self._convertTag('<i>(.*?)<\\/i>', '[i]${match[1]}[/i]', result);
+    result = self._convertTag('<img>(.*?)<\\/img>', '[img]${match[1]}[/img]', result);
+    result = self._convertTag('<color=#(.*?)>(.*?)<\\/color>', '[color=#${match[1]}]${match[2]}[/color]', result);
```
Comment (blurymind) [inline src/js/classes/richTextFormatterBbcode.js:109]: the <color> tag doesnt work in the formatting previewer.
Try `<color=#2ed509>next?</color>`

**Comment by blurymind** [comment]
The rendering of the node preview sometimes doesnt properly format html tags.
Note how the bold tag has been missed.
Screenshot_2020-05-15_18-28-07

Strangely enough it is fine when you preview it in the editor
Screenshot_2020-05-15_18-32-28


(with the caveat color tags dont work at all)

I believe the node preview and the format preview in the node editor used the same function to render this. Why is it getting inconsistent rendering? Are you testing this with all tags? :)

Try with
[NewFile (2).json.zip](https://github.com/YarnSpinnerTool/YarnEditor/files/4635710/NewFile.2.json.zip)

File: src/js/classes/richTextFormatterHtml.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  this.getTagClose = function(tag) {
+    switch (tag) {
+    case 'cmd': return '>>';
+    case 'opt': return '|]]';
+    default: return `</${tag}>`;
+    };
+  };
+
+  this.identifyTag = function(text) {
+    let tag = text.lastIndexOf('<') !== -1 ?
```
Comment (blurymind) [inline src/js/classes/richTextFormatterHtml.js:32]: I am not a big fan of this code duplication. Can we have these two classes share stuff so we dont have to fix it in each every time? :)

Comment (blurymind) [reply src/js/classes/richTextFormatterHtml.js:32]: Maybe have them both extend the same class? 
https://medium.com/@_kamerontanseli/quick-guide-to-using-interfaces-with-javascript-5a557f635e11

File: src/index.html (hunk 1 above)
Comment (blurymind) [reply src/index.html:285]: How come your editor has formatted it on this PR, but hasnt on your other PR? You are sort of creating merge conflicts for yourself

Comment (daviddq) [reply src/index.html:285]: It wasn't my editor. The file was indented with "tabs" and I converted to "spaces" to match the indentation we use for .js files.

File: src/js/classes/richTextFormatterBbcode.js (hunk 4 above)
Comment (daviddq) [reply src/js/classes/richTextFormatterBbcode.js:109]: This is super weird: while developing on Windows, the < and > arrived as &lt; and &gt; respectively. But now I'm on Mac and they arrive as < and >... I need to take a deeper look at this.

File: src/js/classes/richTextFormatterHtml.js (hunk 5 above)
Comment (daviddq) [reply src/js/classes/richTextFormatterHtml.js:32]: I'm aware of the problems of duplicating code, but also aware of the problems of sharing code between different implementations of a single interface. This leads to soon one implementation conditioning and interfering with the other. This is something I wanted to avoid for now.

Let's have it fully working first and then there will be time to promote some common code to a base class.

**Comment by daviddq** [comment]
> Something on this PR is causing the node links menu to not render properly sometimes
> Screenshot_2020-05-15_18-17-52
> 
> I have 5 nodes and none of them is listed

Yes... that's the result of your proposal on a previous PR for replacing:

`<span class="title" onmouseover="app.ui.openNodeListMenu('link')">Add Link</span>`

by

`<span class="title" onclick="app.ui.openNodeListMenu('link')">Add Link</span>`

Now the LinksMenu is only built if you click on "Add Link".

To fix it we can use both "onclick" (for mobile) and "onmouseover" (for desktop). Any other ideas?

**Comment by blurymind** [comment]
> > Something on this PR is causing the node links menu to not render properly sometimes
> > Screenshot_2020-05-15_18-17-52
> > I have 5 nodes and none of them is listed
> 
> Yes... that's the result of your proposal on a previous PR for replacing:
> 
> `<span class="title" onmouseover="app.ui.openNodeListMenu('link')">Add Link</span>`
> 
> by
> 
> `<span class="title" onclick="app.ui.openNodeListMenu('link')">Add Link</span>`
> 
> Now the LinksMenu is only built if you click on "Add Link".
> 
> To fix it we can use both "onclick" (for mobile) and "onmouseover" (for desktop). Any other ideas?

sorry I missed that last time, should have noticed it then.

Lets go with both :)

File: src/js/classes/richTextFormatterHtml.js (hunk 5 above)
Comment (blurymind) [reply src/js/classes/richTextFormatterHtml.js:32]: ok that sounds like a good plan :)

**Comment by daviddq** [comment]
Done!

**Comment by blurymind** [comment]
It looks good :)

There are a couple of regressions with the colour picker.

1. Changing colour from the recent ones in the palette no longer updates it in the preview
Peek 2020-05-17 19-46

2. right clicking on a highlighted colour hex no longer brings up the picker to let you change it.

try doing this on your PR:
Here is a gif of how it works on master
https://yarnspinnertool.github.io/YarnEditor/
Peek 2020-05-17 19-45
then try doing it on master, where it works

**Comment by daviddq** [comment]
Fixed! Reading at spectrum-colorpicker docs I could simplify the way we handle the events.

**Comment by blurymind** [comment]
> Fixed! Reading at spectrum-colorpicker docs I could simplify the way we handle the events.

Thank you for fixing the regressions. It works, but it is still somewhat strange. When you update a colour with the picker, its string does not update until you click somewhere on the editor to unselect it. 
Peek 2020-05-18 19-12

On master the sting instantly updates, no extra click required

**Comment by daviddq** [comment]
Weird as hell! I can't reproduce it on my side. Are you testing on Windows or on Linux? In past PRs I found different behaviors between Windows and Mac and managed to solve them. Maybe it's a new case.

**Comment by blurymind** [comment]
> Weird as hell! I can't reproduce it on my side. Are you testing on Windows or on Linux? In past PRs I found different behaviors between Windows and Mac and managed to solve them. Maybe it's a new case.

on linux.. clearing the cache fixed that. Disregard the comment :) I will test some more and get it merged if theres nothing else

**Comment by blurymind** [comment]
I noticed that it happens more if you select palette preset colour 2 times or more before dismissing the picker

**Comment by blurymind** [comment]
Ah it is strange indeed, why is there a timeout now. I need to have a look at what exacly changed in this pr for the picker

**Comment by daviddq** [comment]
I noticed some "glitch" on the color picker, but it's not consistent and it's hard to reproduce. I'll try to guess the pattern. Thanks for the extra info 😉

File: src/js/classes/app.js [hunk 6]
```diff
@@ -703,64 +644,35 @@ export var App = function(name, version) {
           '#ead1dc',
         ],
       ],
-      change: function(color) {
-        if ($('#colorPicker-container').is(':visible')) {
-          self.applyPickerColorEditor(color);
-          $('#colorPicker').spectrum.set(color.toHexString());
-        }
+      move: function(color) {
```
Comment (blurymind) [inline src/js/classes/app.js:647]: why was this changed? :) was it the commit to get rid of some warning

Comment (daviddq) [reply src/js/classes/app.js:647]: The warning said "$('#colorPicker').spectrum.set" is not a function, so I called "set" only if it existed. But then I found out that it was a problem in the order the color picker was shown and reverted the change.

Comment (daviddq) [reply src/js/classes/app.js:647]: But I'm checking the very same function... something smells bad there.

File: src/js/classes/app.js [hunk 7]
```diff
@@ -703,64 +644,35 @@ export var App = function(name, version) {
...
+            if (self.getTagBeforeCursor() === tag.Start) {
+              tag.Completion && self.insertTextAtCursor(tag.Completion);
+              tag.Offset && self.moveEditCursor(tag.Offset);
+              tag.Func && tag.Func();
             }
-          }, 200);
-          return;
-        }
+          });
+        }, 200);
```
Comment (blurymind) [inline src/js/classes/app.js:675]: can you reduce this time out to 10, it seems to make the delay less noticeable

Comment (daviddq) [reply src/js/classes/app.js:675]: That will prevent the autocompletion from recognize [img] and  because it go always for [i] and <i>

File: testFiles/bbcodeTags.yarn [hunk 8]
```diff
@@ -0,0 +1,25 @@
+title: Start
```
Comment (blurymind) [inline testFiles/bbcodeTags.yarn:1]: thanks for this :) :+1:

File: src/js/classes/app.js (hunk 7 above)
Comment (blurymind) [reply src/js/classes/app.js:675]: ah I see. I never liked timeouts, wish we didnt have to use it

File: src/js/classes/app.js (hunk 6 above)
Comment (blurymind) [reply src/js/classes/app.js:647]: it's even more noticeable when you add a new color tag
Peek 2020-05-18 21-10

**Comment by blurymind** [comment]
I found another one, sorry :)
If you call the picker and dismiss it without selecting a colour, you end up with an empty hex and a broken tag
Peek 2020-05-18 21-18
Ideally it should apply the colour its displaying when you dismiss it

**Comment by daviddq** [comment]
No worries, man. The more we find the more we will fix 😉

File: src/js/classes/app.js (hunk 6 above)
Comment (blurymind) [reply src/js/classes/app.js:647]: to clarify this a bit, in the above gif it is not a delay. It updates the colour when I click again - that could be something to do with the input refactor you have on here

**Comment by daviddq** [comment]
The "update color after click" issue should be fixed now. Please, confirm.

**Comment by daviddq** [comment]
The "no color on dismiss" issue should be fixed now. Please, confirm.

**Comment by blurymind** [comment]
thank you for fixing these :) I will give it another test in my lunch break and merge if there's no regressions

**Comment by blurymind** [comment]
both fixed now! 
Thank you for working on this and for the patience :+1: 

Merging now :tada:

**Comment by blurymind** [comment]
@daviddq sorry I missed this, but there are a few problems with this feature:

- when playtesting this bit `<i><u><color=#e900ff>italics, underlined and pink</color></u></i> ` it doesnt get formatted properly
- Playtesting with html tags, the `< ` symbol constantly appears then disappears during text scrolling
- Playtesting while holding the Z key no longer makes the scrolling faster (possibly a regression from another commit) - the idea is that the scrolling speed is normal when not holding the key  and its fast when holding the key

**Comment by blurymind** [comment]
Whats even worse is node autocreation seems completely broken now :cry: 
https://github.com/YarnSpinnerTool/YarnEditor/issues/187#issuecomment-634671720

**Comment by blurymind** [comment]
I fixed the text scrolling speed issue with the z button :)
